### PR TITLE
[Networking] Migrates topology management to GossipSub

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR simplifies the topology management at Flow codebase two only fully-connected, and empty-list ones. The former is utilized on every staked Flow role and allows a node to connect to every other node on the networking layer. The latter is used on unstaked Access Node's public network where participants' identities are not known apriori. 

Switching to the fully-connected topology simplifies the topology management at the Flow level and leverages GossipSub's native topology management (through peer scoring). In order to be resilient against attacks such as censorship and eclipse, the GossipSub relies on the random eager gossip to nodes out of the local mesh, and having a fully-connected topology guarantees the odds of gossip routing literally circumvents the eclipsing or censoring malicious nodes.

This PR also adds unit testing covering the integration of `PeerManager` and `Libp2pConnector` on fully-connected topology, i.e., a node connects to all other nodes based on the id provider, and trims connections to those that are evicted, ejected, or unstaked.

### CODE REVIEW

**Review by vishalchangrani — APPROVED** [review]
lgtm

**Review by kc1116 — COMMENTED** [review]
Looks good, left a few small comments.

File: network/p2p/peerManager_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	// after a forced update all other nodes must be added to the peer store of this node.
+	require.Len(t, thisNode.Host().Network().Peers(), count-1)
+	// after a forced update there must be a connection between this node and other nodes
+	connectedToAll(t, thisNode.Host(), idTranslator, othersId.NodeIDs())
+
+	// kicks one node out of the othersIds; this imitates evicting, ejecting, or unstaking a node
+	evictedId := othersId[0] // evicted one
+	othersId = othersId[1:]  // updates otherIds list
+	peerManager.ForceUpdatePeers()
+	time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/peerManager_integration_test.go:76]: No, it is just to ensure that we are waiting enough till peer manager establishes a connection to all peers. This is done asynchronously at the LibP2P level so we do not have a synchronous blocking hook to deterministically progress when connections are established. We typically do such sleep on [similar networking primitives](https://github.com/onflow/flow-go/blob/50011e2a2e9b1877a3bbf280d38d60097d31198d/network/test/meshengine_test.go#L388-L391) that their lifecycle is asynchronous to the test process.

File: network/p2p/peerManager_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	count := 5
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// create nodes
+	nodes, identities := nodesFixture(t, ctx, unittest.IdentifierFixture(), "test_peer_manager", count)
+	defer stopNodes(t, nodes)
+
+	thisNode := nodes[0]
+	othersId := identities[1:]
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/peerManager_integration_test.go]: https://github.com/onflow/flow-go/pull/2875/commits/035f6dbe620cca8a2bbf26d6c811821b7b61e070

File: network/test/middleware_test.go [hunk 3]
```diff
@@ -97,10 +93,10 @@ func (m *MiddlewareTestSuite) SetupTest() {
 	peerChannel := make(chan string)
 	ob := tagsObserver{
 		tags: peerChannel,
-		log:  logger,
+		log:  unittest.Logger(),
 	}
 
-	m.ids, m.mws, obs, m.providers = GenerateIDsAndMiddlewares(m.T(), m.size, logger, unittest.NetworkCodec())
+	m.ids, m.mws, obs, m.providers = GenerateIDsAndMiddlewares(m.T(), m.size, unittest.Logger(), unittest.NetworkCodec())
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/test/middleware_test.go]: https://github.com/onflow/flow-go/pull/2875/commits/3177f848057efb9ed2c17c7b8424db63eb191b07

File: network/test/middleware_test.go [hunk 4]
```diff
@@ -97,10 +93,10 @@ func (m *MiddlewareTestSuite) SetupTest() {
 	peerChannel := make(chan string)
 	ob := tagsObserver{
 		tags: peerChannel,
-		log:  logger,
+		log:  unittest.Logger(),
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/test/middleware_test.go]: https://github.com/onflow/flow-go/pull/2875/commits/3177f848057efb9ed2c17c7b8424db63eb191b07

**Review by kc1116 — APPROVED** [review]

File: network/p2p/middleware.go [hunk 5]
```diff
@@ -211,18 +211,6 @@ func (m *Middleware) NewPingService(pingProtocol protocol.ID, provider network.P
 	return NewPingService(m.libP2PNode.Host(), pingProtocol, m.log, provider)
 }
 
-// topologyPeers callback used by the peer manager to get the list of peer ID's
-// which this node should be directly connected to as peers.
-// No errors are expected during normal operation.
-func (m *Middleware) topologyPeers() (peer.IDSlice, error) {
```
[the remark this answers is not in the document]
Comment (yhassanzadeh13) [reply network/p2p/middleware.go:217]: https://github.com/onflow/flow-go/pull/2875/commits/b5969d2069cd21e165fa908b693c4c8cc60d7265

**Pull request merged**